src/ui/grid_layout.h

```diff
@@ -0,0 +1,171 @@
+#pragma once
+#include "dashboard.h"
+#include <ranges>
+
+namespace DigitizerUi {
+
+enum class GridArrangement {
+    Horizontal,
+    Vertical,
+    Tiles,
+    Free,
+};
+
+class GridLayout {
+public:
+    constexpr static inline auto grid_width  = 16u;
+    constexpr static inline auto grid_height = 16u;
+    constexpr static inline auto max_plots   = 16u;
+    struct dirty_guard {
+        bool &flag;
+        dirty_guard(bool &flag) noexcept
+            : flag(flag) {}
+        ~dirty_guard() { flag = false; }
+    };
+
+public:
+    void ArrangePlots(std::span<Dashboard::Plot> plots) noexcept {
+        dirty_guard guard{ dirty };
+        if (!plots.size()) return;
+
+        if (!dirty)
+            return MaintainLayout(plots);
+
+        switch (m_arrangement) {
+        case DigitizerUi::GridArrangement::Horizontal:
+            return RearrangePlotsHorizontal(plots);
+        case DigitizerUi::GridArrangement::Vertical:
+            return RearrangePlotsVertical(plots);
+        case DigitizerUi::GridArrangement::Tiles:
+            return RearrangePlotsTiles(plots);
+        case DigitizerUi::GridArrangement::Free:
+            break;
+        default:
+            break;
+        }
+    }
+    void SetArrangement(GridArrangement arrangement) noexcept {
+        dirty         = true;
+        m_arrangement = arrangement;
+    }
+    auto Arrangement() const noexcept { return m_arrangement; }
+    void Invalidate() noexcept { dirty = true; }
+
+private:
+    void MaintainLayout(std::span<Dashboard::Plot> plots) {
+        switch (m_arrangement) {
+        case DigitizerUi::GridArrangement::Horizontal: {
+            uint32_t snap = grid_width;
+            for (auto it = plots.rbegin(); it != plots.rend(); it++) {
+                auto& i = *it;
+                i.rect.h = grid_height;
+                i.rect.y = 0;
+                i.rect.w = snap - i.rect.x;
+                snap     = i.rect.x;
+            }
+            break;
+        }
+        case DigitizerUi::GridArrangement::Vertical: {
+            uint32_t snap = grid_width;
+            for (auto it = plots.rbegin(); it != plots.rend(); it++) {
+                auto& i = *it;
+                i.rect.w = grid_height;
```

Comment:
could be merged to `it->rect.w = ...`  similar occurrences below.

Maybe the iterator `it` could get a better name? 'plot'?